### PR TITLE
Format word vectors

### PR DESCRIPTION
**Pull request by arynh**
Resolves issue: final formatting for handoff

* Make an object for each clustering method
* Add example usages in jupyter notebooks
* good docs

### CODE REVIEW

File: processing/WordVectors.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+import numpy
+import spacy
+from gensim.parsing.preprocessing import preprocess_string, remove_stopwords
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import StandardScaler
+
+
+def get_words(text):
+    """
+    Source: <@FIXME:/somewhere/>
```
Comment (arynh) [inline processing/WordVectors.py]: fix it (add reference to Jackie(/prashant/eugenia/lam/josh/aryn)'s code or import from it)

File: processing/WordVectors.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    for word in wordlist:
+        # only get words (no digits)
+        if not word.isdigit() and not re.match(r"[^\w]", word):
+            clean_list.append(word)
+    return " ".join(clean_list)
+
+
+class WordVectors:
+    """
+    @FIXME: class to represent word vectors. more stuff here
```
Comment (arynh) [inline processing/WordVectors.py]: why haven't you written this yet

File: processing/WordVectors.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    """
+
+    dataset = []  # list of dictionaries with "summary" and "_id" fields
+    # @FIXME: or perhaps we ask for a more general input format?
```
Comment (arynh) [inline processing/WordVectors.py]: make this more general; perhaps use `{"summary" : <summary>, "theme" : <theme> 🔢}`

File: processing/WordVectors.py (hunk 2 above)
Comment (arynh) [reply processing/WordVectors.py]: list each method, describe class variables/onmline comments

File: processing/WordVectors.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    samples = 500  # just used for visualization; we don't want to dump all points
+    # onto the plot, as they will overlap and overwhelm
+
+    word_vectors = []
+    word_embedding = None  # word embedding model; pretrained from spacy.io
+
+    X = None
```
Comment (arynh) [inline processing/WordVectors.py]: featyre matrix

File: processing/WordVectors.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    explained_variance = None
```
Comment (arynh) [inline processing/WordVectors.py:46]: fre k ot ot his

File: processing/WordVectors.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+    def __init__(self, data_file: string, samples: int):
```
Comment (arynh) [inline processing/WordVectors.py]: this won't work; also change to data in memory input. samples should have a default value

File: processing/WordVectors.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        vectors for each document and normalizes them.
+        Returns a list of naive document vectors.
+        """
+        self.word_vectors = [
+            self.word_embedding(remove_stopwords(get_words(doc["summary"]))).vector
+            for doc in self.dataset[: self.samples - 1]
+        ]
+        return self.word_vectors
+
+    def reduce_word_vectors(self, vectors):
```
Comment (arynh) [inline processing/WordVectors.py]: mayhaps allow other component values; something other than two might e usefule

Comment (arynh) [reply processing/WordVectors.py]: use 2 as defau`l`t

File: processing/WordVectors.py [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        """
+        Perform dimensionality reduction on feature matrix using PCA.
+        vectors: list of word vectors
+        Returns:
+            X: reduced feature matrix, numpy.ndarray Nx2
+            explained_variance: percentage increase in explained variance due
+                to each principal component
+        """
+        X = numpy.vstack(vectors)
+        scaler = StandardScaler()
```
Comment (arynh) [inline processing/WordVectors.py:89]: what is this

File: processing/WordVectors.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        pca.fit_transform(X)
+        self.X = X
+        return X, explained_variance
+
+    def visualize(self, vector_labels):
+        """
+        Plots the reduced matrix of document embeddings using MatPlotLib.
+        Plotted documents are colored according to given labels.
+        Also plots a graph of explained variance vs. dimensions
+        Returns:
```
Comment (arynh) [inline processing/WordVectors.py:101]: `vector_labels` is an input

Comment (arynh) [reply processing/WordVectors.py:101]: miao return the figure

Comment (arynh) [reply processing/WordVectors.py:101]: 😼

**Review by arynh — COMMENTED** [review]
pls change

File: processing/WordVectors.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        text_key = "summary"
+        if not self.dataset[key]:
+            key = "text"
+        self.word_vectors = [
+            self.word_embedding(remove_stopwords(get_words(doc[text_key]))).vector
+            for doc in self.dataset[: self.samples - 1]
+        ]
+        return self.word_vectors
+
+    def reduce_word_vectors(self, vectors: list, pca_components=2):
```
Comment (JoshBurke) [inline processing/WordVectors.py]: self.word_vectors can be a default value for vectors

File: processing/WordVectors.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        self.X = X
+        return X, explained_variance
+
+    def visualize(self, vector_labels: list):
+        """
+        Plots the reduced matrix of document embeddings using MatPlotLib.
+        Plotted documents are colored according to given labels.
+        Also plots a graph of explained variance vs. dimensions
+        Parameters:
+            vector_labels: list of labels for each document
```
Comment (JoshBurke) [inline processing/WordVectors.py]: wat r these tho

File: processing/WordVectors.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        Returns:
+            (fig, ax): the figure object and axis object which describes the
+                figure. This is returned so it can be plotted or saved.
+        """
+        labels = numpy.array(vector_labels)
+        fig, ax = plt.subplots(2)
+        labelset = set(labels)
+        for label in labelset:
+            ax[0].plot(
+                self.X[labels == label, 0],
```
Comment (JoshBurke) [inline processing/WordVectors.py:110]: ah yes but why?

**Comment by arynh** [comment]
make the flow consistent between how each method works. Does is operate on given data, or on the data in the class variable? Also fix the broken labels. Something is missing.
